lib/spack/spack/test/bindist.py

```diff
@@ -569,3 +575,40 @@ def test_built_spec_cache(tmpdir,
     margs = mparser.parse_args(
         ['rm', '--scope', 'site', 'test-mirror-rel'])
     mirror.mirror(mparser, margs)
+
+
+def test_spec_needs_rebuild(install_mockery_mutable_config, mock_packages,
+                            mock_fetch, monkeypatch, tmpdir):
+    """Make sure needs_rebuild properly comares remote full_hash
+    against locally computed one, avoiding unnecessary rebuilds"""
+
+    # Create a temp mirror directory for buildcache usage
+    mirror_dir = tmpdir.join('mirror_dir')
+    mirror_url = 'file://{0}'.format(mirror_dir.strpath)
+
+    mirror_cmd('add', 'test-mirror', mirror_url)
+
+    s = Spec('libdwarf').concretized()
+
+    # Install a package
+    install_cmd(s.name)
+
+    # Put installed package in the buildcache
+    buildcache_cmd('create', '-u', '-a', '-d', mirror_dir.strpath, s.name)
+
+    rebuild = bindist.needs_rebuild(s, mirror_url, rebuild_on_errors=True)
+
+    assert not rebuild
+
+    # Now monkey patch Spec to change the full hash on the package
+    def fake_full_hash(spec):
+        print('fake_full_hash')
+        return 'tal4c7h4z0gqmixb1eqa92mjoybxn5l6'
+    monkeypatch.setattr(spack.spec.Spec, 'full_hash', fake_full_hash)
+
+    rebuild = bindist.needs_rebuild(s, mirror_url, rebuild_on_errors=True)
+
+    assert rebuild
+
+    uninstall_cmd('-y', s.name)
+    mirror_cmd('rm', 'test-mirror')
```

Comment:
I think the `install_mockery` and `mutable_config` fixtures will take care of this. If there aren't existing fixtures that can handle this, we need to add new ones. This code will not be run if the test fails.

Reply:
It seems the fixture the test is using now, `install_mockery_mutable_config`, does in fact take care of this, so I removed those lines.